### DISCUSSION TITLE
bug? github action runs and artifacts retrieved by REST API don't match the website

### DISCUSSION THREAD

**Discussion by mattvenn** [community/community#28880] [Apps, API and Webhooks]
I think I've found a bug on the REST API to fetch github actions artifacts.

If I go to this page: https://github.com/wokwi/tiny-tapeout-test-simple/actions
I see 2 actions results. The top one is the most recent, with run id 2823947105. Clicking on that and then looking at the artifact url for downloading the zip gives: https://github.com/wokwi/tiny-tapeout-test-simple/suites/7729232355/artifacts/323987363

Now I use a Python program accessing the repo via the REST API. Here are the 2 actions results

```
ipdb> data['artifacts'][0]['workflow_run']['id'], data['artifacts'][0]['archive_download_url'], data['artifacts'][0]['created_at']
(2823942817, 'https://api.github.com/repos/wokwi/tiny-tapeout-test-simple/actions/artifacts/323988021/zip', '2022-08-09T08:50:58Z')
ipdb> data['artifacts'][1]['workflow_run']['id'], data['artifacts'][1]['archive_download_url'], data['artifacts'][1]['created_at']
(2823947105, 'https://api.github.com/repos/wokwi/tiny-tapeout-test-simple/actions/artifacts/323987363/zip', '2022-08-09T08:50:13Z')
```

The first one should be the most recent, and it does have the most recent time stamp, but the job and download url are the older run.
The second one has the older timestamp, but the correct new results.

I have 4 other repos I am testing this on, and all the others return the correct results. Should I open an issue - and if so on what repository?

**Comment by airtower-luna** [comment]
I can't really say if it's wrong: The workflows were started very close together, and the first one took more time, so it'd be plausible the times are when the artifacts were finalized.

If you want to file a bug report the support form is here: https://support.github.com/contact/bug-report

**Reply by mattvenn** [reply]
Aha great clue. I was just assuming they would take the same time but of course that's not true. So if I want to get the last build I need to sort on the job started time.

**Reply by mattvenn** [reply]
What I said above is wrong. The first item on the list is the one that started and ended later, but it belongs to the previous commit. So in this case the 2nd item belongs to the later commit. Which makes sense if github took longer to assign the job to a runner for the first commit than the second.


So I think I need to first fetch the commits, which will be in date order, then sort the action commit by the commit history, then select the most recent successful one.

Probably not a bug, but definitely easy to not get the expected result.

**Reply by mattvenn** [reply]
This works, if a bit ugly
```
111         # the latest artifact isn't necessarily the one related to the latest commit, as github                                                     
112         # could have taken longer to process an older commit than a newer one.                                                                      
113         download_url = None                                                                                                                         
114         for commit in commits:                                                                                                                      
115             for artifact in data['artifacts']:                                                                                                      
116                 commit_sha = commit['sha']                                                                                                          
117                 action_sha = artifact['workflow_run']['head_sha']                                                                                   
118                 logging.debug("commit {} action {}".format(commit_sha, action_sha))                                                                 
119                 if commit_sha == action_sha:                                                                                                        
120                     download_url = artifact['archive_download_url']
121                     break                                                                                                                           
122             if download_url is not None:
123                 break                                                                                                                               
```

**Reply by airtower-luna** [reply]
Possible catch: Depending on your workflow setup (and whether you use "re-run") there may be multiple runs for the same commit. :sweat_smile:
